### PR TITLE
Separate employee and customer login routes

### PR DESCRIPTION
**Pull request by sudo0rw3ll**
Hello team,
In this ticket lots of changes have been made. I will list and explain each one of them.

1. User class in model package is changed to be @MappedSuperclass since Employee and Customer
have lots of fields in common but also some of the fields are unique for each entity. 

Strategy is set to TABLE_PER_CLASS since there is no need for User table since it's abstract. Because of this strategy, generation type of primary identifiers is also changed from IDENTITY to TABLE due to table creation strategy change.

2. Mapper for Employee is created and it contains methods which perform transformations from Employee object to Dtos and vice versa. Customer mapper has been change along with User mapper since they invoked some methods which are not available anymore due to User class changes. 

To be more specific, I am speaking of methods .setPostition(arg) which is not present in User class since this field has no meaning for User and it's been moved to Employee, .setPermissions(listArg)  because User class has no permission set anymore because Permission entity cannot establish relation with MappedSuperclass. Speaking of this, Permission class has also been change to contain Sets of Employees and Customers to maintain relations.

3. Repository for Employee has been added and methods from UserRepository have been COPIED (not modified) into this file.
Please leave comments for adjusting bussines logic of SQL queries.

4. Created Employee controller with routes prefixed with /employee because I wanted to test initial changes. This can be easily changed to /user. 

5. Created Employee service which extends UserDetailsService in order to enable login functionality. 

6. Implemented EmployeeService with small changes in source code.

7. Changed Customer service to extend UserDetailsService in order to enable separate login functionality.
Before, there was only User entity and User Service held all responsibility when it comes to dealing with UserDetails. Now, situation is different because we have separated Employee and Customer entities from User. 

Before, lookup was much easier because AuthenticationManager had to call userService.loadByUsername() method which would query the User table to check if Employee or Customer with certain email is present. 

Now, when we have separate tables we cannot easily call one loadByUsername() method because we have to check if user that's logging in is Employee or Customer. That's why both Employee and Customer services extend UserDetailsService and implement loadByUsername() method. 

In security config authentication manager was relying on userService to invoke loadByUsername. I had to write CustomAuthProvider which would try to extract UserDetails by invoking customerService.loadByUsername() and if this method threw UsernameNotFoundException that would mean that user with this email is not present inside the table and that I should check in Employee table.

8. TESTS
I am not sure how well tests have been changed since I adapted the code to use Components which were related to Employee and not to User. Inside UserControllerSteps I have made same changes and changed routes from /user to /employee because of the controller.

Tests were successfull on my machine but I would suggest @savatarko to check them and comment If any changes are needed.

Best,
Vid

### CODE REVIEW

File: user-service/src/main/java/rs/edu/raf/banka1/services/EmployeeServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+            }
+        }
+
+        this.employeeRepository.save(employee);
+
+        return true;
+    }
+
+    @Override
+    public Boolean sendResetPasswordEmail(String email) {
```
Comment (ijevtic) [inline user-service/src/main/java/rs/edu/raf/banka1/services/EmployeeServiceImpl.java:196]: koliko se secam uradjeno je ovo i za customera?

Comment (ijevtic) [reply user-service/src/main/java/rs/edu/raf/banka1/services/EmployeeServiceImpl.java:196]: ako nije, @MarkoNik je trebalo da to uradi

File: user-service/src/main/java/rs/edu/raf/banka1/controller/EmployeeController.java [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+                            schema = @Schema(implementation = EditEmployeeDto.class))}),
+            @ApiResponse(responseCode = "404", description = "User not found"),
+            @ApiResponse(responseCode = "500", description = "Internal server error")
+    })
+    public ResponseEntity<Boolean> editUser(@RequestBody EditEmployeeDto editUserRequest) {
+        boolean edited = employeeService.editUser(editUserRequest);
+        return new ResponseEntity<>(edited, edited ? HttpStatus.OK : HttpStatus.NOT_FOUND);
+    }
+
+    @PostMapping(value = "permission/{userId}")
```
[the remark this answers is not in the document]
Comment (sudo0rw3ll) [reply user-service/src/main/java/rs/edu/raf/banka1/controller/EmployeeController.java:193]: Iskreno, rute nisam gledao jer sam ih kopriao iz users controller-a kada sam pravio employee kontroler. 
Uradicu izmene, ovo mi je totalno promaklo.

File: user-service/src/main/java/rs/edu/raf/banka1/controller/EmployeeController.java [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+                            schema = @Schema(implementation = Boolean.class))}),
+            @ApiResponse(responseCode = "404", description = "User not found"),
+            @ApiResponse(responseCode = "500", description = "Internal server error")
+    })
+    public ResponseEntity<Boolean> deleteUser(@PathVariable Long id) {
+        Boolean deleted = employeeService.deleteUser(id);
+        return new ResponseEntity<>(deleted, deleted ? HttpStatus.OK : HttpStatus.NOT_FOUND);
+    }
+
+    @GetMapping(value = "/permissions/userId/{userId}", produces = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (sudo0rw3ll) [reply user-service/src/main/java/rs/edu/raf/banka1/controller/EmployeeController.java:224]: Slazem se

File: user-service/src/main/java/rs/edu/raf/banka1/dtos/employee/CreateEmployeeDto.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
+package rs.edu.raf.banka1.dtos.employee;
+
+public class CreateEmployeeDto {
```
[the remark this answers is not in the document]
Comment (sudo0rw3ll) [reply user-service/src/main/java/rs/edu/raf/banka1/dtos/employee/CreateEmployeeDto.java:3]: Nisam koristio anotacije jer lombok generalno ne koristim vec sam rucno napravio getere i setere za DTO.

File: user-service/src/main/java/rs/edu/raf/banka1/model/Customer.java [hunk 5]
```diff
@@ -65,4 +66,11 @@ public class Customer extends User {
     )
     private List<BankAccount> accountIds;
 
+    @ManyToMany(fetch = FetchType.EAGER)
```
[the remark this answers is not in the document]
Comment (sudo0rw3ll) [reply user-service/src/main/java/rs/edu/raf/banka1/model/Customer.java:69]: Slazem se, ovo nisam ni primetio prilikom refactor-a. Takve su bile definicije join tabela koje sam samo preneo.

[the remark this answers is not in the document]
Comment (sudo0rw3ll) [reply user-service/src/main/java/rs/edu/raf/banka1/model/Customer.java:69]: Bice izmenjeno

File: user-service/src/main/java/rs/edu/raf/banka1/bootstrap/BootstrapData.java [hunk 6]
```diff
@@ -49,26 +50,24 @@ public void run(String... args) throws Exception {
...
+        employeeRepository.save(user1);
 
-        User client = new User();
+        Customer client = new Customer();
         client.setEmail("client@gmail.com");
         client.setPassword(passwordEncoder.encode("client"));
         client.setFirstName("Client");
         client.setActive(true);
-        user1.setPosition("employee");
+//        user1.setPosition("employee");
```
[the remark this answers is not in the document]
Comment (sudo0rw3ll) [reply user-service/src/main/java/rs/edu/raf/banka1/bootstrap/BootstrapData.java:68]: Pre izmene, i client i 2 user-a su bili tipa User. Kada sam uradio izmene, User vise nije imao mogucnost setPosition() vec je to bilo moguce samo u Employee pa je iz toga razloga zakomentarisano.

[the remark this answers is not in the document]
Comment (Kovelja009) [reply user-service/src/main/java/rs/edu/raf/banka1/bootstrap/BootstrapData.java:68]: Imaćemo ovde problema, jer Andreja treba da doda aktuare, a kada sam poslednji put video ona se oslanjala na User klasu, pa iz nje to generisala, mora samo da bacim pogled da li je menjala ili ne.

File: user-service/src/main/java/rs/edu/raf/banka1/dtos/employee/CreateEmployeeDto.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kovelja009) [reply user-service/src/main/java/rs/edu/raf/banka1/dtos/employee/CreateEmployeeDto.java:3]: Vidou, je l on ovo tebe potcenjuje?? Dok getter i setter ne prođu  kroz tvoje prste...

File: user-service/src/main/java/rs/edu/raf/banka1/controller/EmployeeController.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+            @ApiResponse(responseCode = "404", description = "Bad request"),
+            @ApiResponse(responseCode = "500", description = "Internal server error")
+    })
+    public ResponseEntity<Boolean> changePermissions(@PathVariable Long userId,
+                                                     @RequestBody ModifyPermissionsRequest request) {
+        boolean changed = employeeService.modifyUserPermissions(request, userId);
+        return new ResponseEntity<>(changed, changed ? HttpStatus.OK : HttpStatus.NOT_FOUND);
+    }
+
+    @DeleteMapping(value = "/remove/{id}")
```
[the remark this answers is not in the document]
Comment (Kovelja009) [reply user-service/src/main/java/rs/edu/raf/banka1/controller/EmployeeController.java:213]: Usput svaka čast za refactor.

[the remark this answers is not in the document]
Comment (sudo0rw3ll) [reply user-service/src/main/java/rs/edu/raf/banka1/controller/EmployeeController.java:213]: Hvala Vanjo! Koliko sam ja skapirao, stvari vezane za customer-a se nalaze u tom controller-u i service-u.

**Comment by Kovelja009** [comment]
Ovde bi trebalo da obratimo pažnju da će potencijalno trebati da dodamo stvari za Customer-a.

**Review by Kovelja009 — APPROVED** [review]

**Review by savatarko — APPROVED** [review]
sto se tice mene okej je, koliko vidim testovi su promenjeni samo sa user na employee

**Pull request merged**